### ISSUE TITLE
Unexpected WordPress.CSRF.NonceVerification.NoNonceVerification warnings

### ISSUE THREAD

**Issue by paulschreiber** [WordPress-Coding-Standards/WordPress-Coding-Standards#1172]
After updating to PHPCS 3.1 and WPCS 0.13.1, I'm seeing a lot of WordPress.CSRF.NonceVerification.NoNonceVerification errors.

I'm using PHP 5.6.30, which did not change.

Here's a snippet that generates the error:

```php
<?php
class Foo {
	public static function bar() {
		if ( ! isset( $_GET['orderby'] ) ) { // input var okay
			1;
		}
	}
}
```

**Comment by jrfnl** [comment]
Updated from what version ? AFAICS, the NonceVerification sniff has not had any significant changes since WPCS 0.9.0.

The error seems correct to me as you're accessing a request variable without doing a nonce check.

**Comment by paulschreiber** [comment]
Upgraded from PHPCS 2.7.1 and WPCS 0.10.0.

**Comment by connerbw** [comment]
Hi there. I have the same issue as @paulschreiber 

After upgrading from 0.10.0 we get hundreds of what I consider bogus NoNonceVerification warnings.

Code that was not being reported about under 0.10.0 is now coming up with warnings. Our project uses `wp_verify_nonce` `check_admin_referer` and `check_ajax_referer` often we know what these are and how to use them but we will also have code like:

```
	public function isFormSubmission() {

		if ( empty( $_REQUEST['page'] ) ) {
			return false;
		}

		if ( $this::PAGE !== $_REQUEST['page'] ) {
			return false;
		}

		if ( $_SERVER['REQUEST_METHOD'] === 'POST' ) {
			return true;
		}

		if ( count( $_GET ) > 1 ) {
			return true;
		}

		return false;
	}
```

The new sniffer rules complain four times for the above function. 

We will be evaluating where we should be making changes and where we we should be using `@codingStandardsIgnore` in the upcoming weeks but `\WordPress\Sniffs\CSRF\NonceVerificationSniff` should be improved to make better use of `\WordPress\Sniff::is_assignment` and `\WordPress\Sniff::is_only_sanitized`.

I think that `isset()` `empty()`, `count()` on requests should not require a nonce. 

I think it's worth discussing whether or not `===` or `!==` on a request should require a nonce. 

I feel only when request values are assigned, or passed to functions, then should we worry about a nonce.

Thank you for your consideration.

**Comment by JDGrimes** [comment]
Here are some examples that demonstrate why a blanket rule for the suggested cases is not a good idea:

```php
if ( ! empty( $_GET['somevar'] ) ) {
    somefunc_that_should_be_protected();
}
```

```php
if ( 'some_value' === $_GET['somevar'] ) {
    somefunc_that_should_be_protected();
}
```

Any time that you access request vars in any way, you are handling a request in some way. Your code is taking into account that a request has been made, and therefore you need to ensure that what it does as a result of that is not something that needs to be protected by an nonce. Just because you are not using the input value does not mean that your code is not responding to the request.

There may be something that we can do here to better when to issue a warning as opposed to an error, but I don't think the sniff should be completely silent in 99+% of cases.

Just detecting a function call doesn't work either, consider this:

 ```php
function is_some_request() {
    if ( ! empty( $_GET['somevar'] ) ) {
        return true;
    }
}

// Elsewhere...

if ( is_some_request() ) {
    somefunc_that_should_be_protected();
}
```

If you follow certain patterns in your code, you may not need some cases flagged because they will always be sage, but just assuming that your code is going to follow the patterns and be safe is not a good idea.

**Comment by connerbw** [comment]
Thanks for your reply. I will have to think about it. 

It's not that you aren't right, it's that by updating my dependency I have added hours of refactoring to my backlog of work that go to the bottom of my priorities. I'm having a hard time understanding what changed between [0.10.0 and now](https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/blob/0.10.0/WordPress/Sniffs/CSRF/NonceVerificationSniff.php) to trigger hundreds of new warnings.

**Comment by GaryJones** [comment]
@connerbw It's not going to help with your general issue, but a couple of points:

- If your project is something where you control the PHP setup (as opposed to a publicly distrubuted plugin), then you could use [filter_input](http://php.net/manual/en/function.filter-input.php). It's enabled by default for PHP 5.2.0+, but it can be disabled.
- I'm not sure if your code example is simplified, but I read that as "if a POST request is made, or a GET request with at least one querystring param is made, and the POST data / querystring contains `page` that matches the value in a static property, then consider it to be a form submission.

If that code was powering `example.com/foo`, then it would be trivial to send an external POST request, or a GET request like `example.com/foo?page=foo` (or whatever the static property value is), and have your function return true.

This is precisely the case where you _do_ need a nonce - if your form has a hidden field with a nonce value (use `wp_nonce_field()`), then you can check against this (`wp_verify_nonce()`). It has an option to also add in a referer field (or manually with `wp_referer_field()`). As you can [see](https://github.com/WordPress/WordPress/blob/f4faaa53c7aa3a5ad4ba9dcd6dc5df711c81d114/wp-includes/pluggable.php#L2072-L2094), it takes into account various factors, so is more reliable that this is indeed a form submission and not just a rogue request that happens to meet the trivial requirements.

**Comment by connerbw** [comment]
The  example is simplified. 

```pjp
public function formSubmit() {
	if ( empty( $this->isFormSubmission() ) || empty( current_user_can( 'edit_others_posts' ) ) ) {
		// Don't do anything in this function, bail.
		return;
	}

	// Process form
	// Process form
	if ( 
		isset( $_GET['custom_styles'] ) && 
		$_GET['custom_styles'] === 'yes' && 
		isset( $_POST['your_styles'] ) && 
		check_admin_referer( 'pb-custom-styles' ) ) 
	{
```

Note the many places this bails correctly. ;-)

I would also like to point out that "coding" and "coding standards" are not really the same thing, and as much as I appreciate the free advice I feel the metaphor is being mixed.

Thanks,

**Comment by connerbw** [comment]
I'm not trying to come off snarky, I will refactor, the sniff is good, but I still feel a discussion is warranted.

My points:

I have been running 0.10.0 since the spring, we did a monumental refactor to comply to the standards, this warning never popped up. Updated today, many warnings. What changed? Same question as the original poster I think?

`check_admin_referer` and friends are used, they are just not on the same line . This particular sniff is making a lot of assumptions as to how one should write code. Good points made but is it this a standard or something else?

**Comment by connerbw** [comment]
Ok. I can answer my own question. What changed:

```
        public $errorForSuperGlobals = array( '$_POST', '$_FILE' );
       // ... snip ...

	// If we're still here, no nonce-verification function was found.
	$severity = ( in_array( $instance['content'], $this->errorForSuperGlobals, true ) ) ? 0 : 'warning';
```
The above was in 0.10.0, it's not in the new code.

We refactored for $_POST and $_FILE, everything else gets 0.

**Comment by connerbw** [comment]
This is what I think is happening 

In `\PHP_CodeSniffer\Files\File::addMessage` we see `$severity` as a param and it expects an `int`, later in that same function is :

```
        if ($includeAll === false && $configSeverity > $severity) {
            return false;
        }
```

In 0.10.0 the code would set `$severity = 'warning'` for anything that wasn't `$_POST` or `$_FILE`.

```
$configSeverity = 5;
$severity = 'warning';
if ($configSeverity > $severity) {
    var_dump("DOH!");
}
```

**Comment by GaryJones** [comment]
https://github.com/WordPress-Coding-Standards/WordPress-Coding-Standards/compare/f60bfe5c3aad100b191192ea1d06e515df3d8f27...de8044b277aeb73b9b3de8660071b3c04854ef58#files_bucket is the `0.10.0` vs `0.14.0` comparison.

There are no obvious changes to the called functions in `Sniff.php`.

**Comment by JDGrimes** [comment]
@connerbw Yes, I guess this may have been a bug in the sniff that was inadvertantly fixed.

>This particular sniff is making a lot of assumptions as to how one should write code. Good points made but is this about a standard or something else?

The sniff has to make assumptions to be able to work; phpcs is not designed for complex code analysis, so it can't be used to detect if a function is being called in a different part of the code.

However, we actually see this as a good thing. Why should it take complex analysis to determine if your code is secure? Isn't it much better if anybody can see at a glance whether it is secure or not? That way, a lack of an nonce check is conspicuous and worth investigating, rather than it looking no different from code that is nonce protected, just with the nonce check somewhere else.

The sniff is not about code style, as you point out, and that is why it is part of the `WordPress-Extra` ruleset, and not `WordPress-Core`. So you are welcome to exclude it from your ruleset (though perhaps at your own peril).

If there are safe cases that we can detect better, I'm happy to improve the sniff, but the examples and suggestions presented so far would result in false negatives. For security-related sniffs like this we err on the side of having false positives, rather than false negatives. The risk of a missed security issue outweighs the burden of dealing with non-issues, at least in my mind.

**Comment by connerbw** [comment]
There was a bug in 0.10.0 that would never show (me) warnings because the default severity level of `5 > 'warning'`

I can provide code that proves this bug existed, but since it's now fixed, I think it's pointless? For the sake of this discussion there was, however, a change. This bug suppressed warnings.

I will accept that I now have to re-factor for warnings that I had previously not seen based on the tips provided.

Thanks for the discussion. Good times.

**Comment by JDGrimes** [comment]
Since it has now been determined that there was indeed a bug that was inadvertently fixed, I'm going to close this issue. Ideas about how to reduce false-positives from this sniff are welcome, but let's open new tickets for further discussion.

**Comment by lkraav** [comment]
What's annoying about it, is that `$_REQUEST` can't apparently (as of PHP 7.2) be smoothly handled with `filter_input`.

Also, how does frontend caching play into all this?
